crates/cranelift/src/obj.rs

```diff
@@ -289,80 +394,50 @@ impl<'a> ObjectBuilder<'a> {
         Ok(())
     }
 
-    pub fn finish(&mut self, isa: &dyn TargetIsa) -> Result<()> {
-        self.append_relocations()?;
+    pub fn finish(&mut self) -> Result<()> {
+        // Now that all function symbols are available register all final
+        // relocations between functions.
+        //
+        // FIXME(#3009) once the old backend is removed this loop should be
+        // deleted since there won't be any relocations here.
+        for (r, off) in mem::take(&mut self.relocations) {
+            let symbol = match r.reloc_target {
+                RelocationTarget::UserFunc(index) => self.func_symbols[index],
+                _ => unreachable!("should be handled in `append_func`"),
+            };
+            let (kind, encoding, size) = match r.reloc {
+                Reloc::X86CallPCRel4 => {
+                    (RelocationKind::Relative, RelocationEncoding::X86Branch, 32)
+                }
+                other => unimplemented!("Unimplemented relocation {:?}", other),
+            };
+            self.obj.add_relocation(
+                self.text_section,
+                ObjectRelocation {
+                    offset: off + u64::from(r.offset),
+                    size,
+                    kind,
+                    encoding,
+                    symbol,
+                    addend: r.addend,
+                },
+            )?;
+        }
+
+        // Finish up the text section now that we're done adding functions.
+        const CODE_SECTION_ALIGNMENT: u64 = 0x1000;
```

Comment:
Would it make sense to have some notion of page alignment per platform, or maybe just take the least-common-multiple of those for platforms we support? I'm thinking specifically of macOS/aarch64, where mmap'ing part of a file with only 4K alignment might be problematic with the 16K (?) pages by default.

Reply:
Indeed! This PR is just moving this logic around as opposed to adding it (it's actually been around for quite some time, although it's more important to get right now that we're directly mmap'ing cache artifacts), so I'll leave this as-is for now, but I'll open an issue. The issue with M1 macs was actually *just* fixed in https://github.com/bytecodealliance/wasmtime/pull/3274 as well, so it definitely needs fixing soon!